lib/ack-tracker/batched_ack_tracker.c

```diff
@@ -0,0 +1,261 @@
+  self->pending_ack_record->super.tracker = &self->super;
+  msg->ack_record = (AckRecord *) self->pending_ack_record;
+}
+
+static void
+_track_msg(AckTracker *s, LogMessage *msg)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *)s;
+  _assign_pending_ack_record_to_msg(self, msg);
+  self->pending_ack_record = NULL;
+}
+
+static GList *
+_append_ack_record_to_batch(BatchedAckTracker *self, AckRecord *ack_record)
+{
+  GList *full_batch = NULL;
+  g_mutex_lock(&self->acked_records_lock);
+  {
+    self->acked_records = g_list_prepend(self->acked_records, ack_record);
+    ++self->acked_records_num;
+    if (self->acked_records_num == self->batch_size)
+      {
+        full_batch = self->acked_records;
+        self->acked_records = NULL;
+        self->acked_records_num = 0;
+      }
+  }
+  g_mutex_unlock(&self->acked_records_lock);
+
+  return full_batch;
+}
+
+static void
+_stop_batch_timer(BatchedAckTracker *self)
+{
+  if (iv_timer_registered(&self->batch_timer))
+    iv_timer_unregister(&self->batch_timer);
+}
+
+static void
+_start_batch_timer(BatchedAckTracker *self)
+{
+  if (self->timeout <= 0)
+    return;
+
+  iv_validate_now();
+  self->batch_timer.expires = iv_now;
+  timespec_add_msec(&self->batch_timer.expires, self->timeout);
+
+  iv_timer_register(&self->batch_timer);
+}
+
+static void
+_batch_timeout(gpointer data)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *) data;
+  GList *batch = NULL;
+  g_mutex_lock(&self->acked_records_lock);
+  {
+    batch = self->acked_records;
+    self->acked_records = NULL;
+    self->acked_records_num = 0;
+  }
+  g_mutex_unlock(&self->acked_records_lock);
+  if (batch)
+    {
+      _ack_batch(self, batch);
+    }
+}
+
+static void
+_init_watches(BatchedAckTracker *self)
+{
+  IV_TIMER_INIT(&self->batch_timer);
+  self->batch_timer.cookie = self;
+  self->batch_timer.handler = _batch_timeout;
+}
+
+static void
+_manage_msg_ack(AckTracker *s, LogMessage *msg, AckType ack_type)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *) s;
+  log_source_flow_control_adjust(self->super.source, 1);
+
+  if (ack_type == AT_SUSPENDED)
+    log_source_flow_control_suspend(self->super.source);
+
+  if (ack_type != AT_ABORTED)
+    {
+      if (!iv_timer_registered(&self->batch_timer))
+        _start_batch_timer(self);
+      GList *full_batch = _append_ack_record_to_batch(self, msg->ack_record);
+      if (full_batch)
+        _ack_batch(self, full_batch);
+    }
+  else
+    {
+      _ack_record_free(msg->ack_record);
+    }
+
+  log_msg_unref(msg);
+  log_pipe_unref((LogPipe *)self->super.source);
+}
+
+static void
+_disable_bookmark_saving(AckTracker *self)
+{
+  // TODO -> ??
+}
+
+static void
+_free(AckTracker *s)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *) s;
+  g_mutex_clear(&self->acked_records_lock);
+
+  _stop_batch_timer(self);
+
+  if (self->acked_records)
+    _ack_batch(self, self->acked_records);
```

Comment:
Is it safe to ack in free? I think the ack callback might want to use some variables of the source, but the source itself is freed earlier:
```
void
log_source_free(LogPipe *s)
{
  LogSource *self = (LogSource *) s;

  g_free(self->name);
  g_free(self->stats_id);
  g_free(self->stats_instance);
  log_pipe_detach_expr_node(&self->super);
  log_pipe_free_method(s);

  ack_tracker_factory_unref(self->ack_tracker_factory);
  ack_tracker_free(self->ack_tracker);
  self->ack_tracker = NULL;

  if (G_UNLIKELY(dynamic_window_is_enabled(&self->dynamic_window)))
    _release_dynamic_window(self);
}
```

Reply:
It should not happen as `LogSource` is  a refcounted `LogPipe` and whenever a message is tracked by the acktracker, the refcounter is increased. So `log_source_free` is called _after_ all messages (tracked by the tracker) has been acked.

Reply:
In this case, what is the point of acking in `AckTracker` `free`, when we know, that it will only be called, after everything is acked?

Reply:
@alltilla : connection between `LogSource` and `AckTracker` is a little bit weird: just see how complex the creation of the `AckTracker`. One of the reasons of this complexity is come from the design of the `LogProtos`: they can tell the type of the AckTracker they will need (rltp...) - and sometimes they don't.
How are the aforementioned problems are related? We know that AckTracker has to keep LogSource alive... now.
But if we could make AckTracker part of LogSource - then it would be a different story.

And there is another reason - a more practical: I want to test `BatchedAckTracker` separately from `LogSource`  (your question is absolutely valid as there is no sign of these tests in the code...)